File: app/src/main/java/com/tencent/tcmpp/demo/open/payment/PaymentManager.java
package com.tencent.tcmpp.demo.open.payment;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.text.TextUtils;
import android.util.Log;
import android.widget.Toast;

import com.tencent.tcmpp.demo.R;
import com.tencent.tcmpp.demo.activity.PaymentResultActivity;
import com.tencent.tcmpp.demo.open.login.Login;
import com.tencent.tcmpp.demo.open.login.LoginApi;
import com.tencent.tcmpp.demo.utils.GlobalConfigureUtil;
import com.tencent.tmf.base.api.config.ITMFConfigManager;
import com.tencent.tmf.core.api.TMFServiceManager;
import com.tencent.tmfmini.sdk.launcher.core.IMiniAppContext;
import com.tencent.tmfmini.sdk.launcher.core.proxy.AsyncResult;

import org.json.JSONException;
import org.json.JSONObject;

public class PaymentManager {
    private static final String TAG = "PaymentManager";
    private static PaymentManager instance;
    private final PayApi payApi = new PayApi();
    private PaymentRequest paymentRequest;

    public static PaymentManager g() {
        if (instance == null) {
            instance = new PaymentManager();
        }
        return instance;
    }

    public void startPayment(IMiniAppContext miniAppContext, JSONObject params, AsyncResult result) {
        //do params check
        if (!params.has("prepayId")) {
            //ill params
            JSONObject failRet = new JSONObject();
            try {
                failRet.put("success", false);
            } catch (JSONException ignored) {
            }
            result.onReceiveResult(false, failRet);
            return;
        }
        //mock check order
        if (isMock(miniAppContext.getContext())) {
            JSONObject mockData = new JSONObject();
            try {
                mockData.put("total_fee", 9999);
            } catch (JSONException ignored) {
            }
            checkOrderMock(miniAppContext, mockData, result);
            return;
        }
        //real check order
        checkOrder(miniAppContext, params, result, PayApi.PAY_TYPE_APP);
    }

    public void notifyPaymentCancel() {
        if (null != paymentRequest) {
            JSONObject ret = new JSONObject();
            try {
                ret.put("errMsg", "cancel");
            } catch (JSONException ignored) {
            }
            paymentRequest.result.onReceiveResult(false, ret);
            paymentRequest = null;
        }
    }

    public void showPwdConfirm(Activity activity, int iconSrc, String desc, String model, String modelId) {
        if (null != paymentRequest) {
            AsyncResult result = paymentRequest.result;
            JSONObject checkRet = paymentRequest.params;
            try {
                checkRet.put("appId", paymentRequest.miniAppContext.getMiniAppInfo().appId);
            } catch (JSONException | NullPointerException ignored) {
            }
            showPwdDialog(activity, paymentRequest.payValue, (retCode, msg, dialogInterface) -> {
                if (retCode == 1) {
                    String pwd = ((CustomPayDemo.CustomPayDialog) dialogInterface).getInputText();
                    Log.e(TAG, "onDismiss isComplete=" + pwd);
                    if (!TextUtils.isEmpty(pwd)) {
                        checkPwdAndPay(activity, pwd, checkRet, result, model, modelId);
                    } else {
                        Log.e(TAG, "empty pwd ~");
                        JSONObject ret = new JSONObject();
                        try {
                            ret.put("errMsg", "empty pws");
                        } catch (JSONException ignored) {
                        }
                        result.onReceiveResult(false, ret);
                        activity.finish();
                    }
                    paymentRequest = null;
                }
            }, iconSrc, desc);
        }
    }

    /**
     * STEP 1: check order status
     *
     * @param miniAppContext
     * @param params
     * @param result
     * @param type           0 mini app,1 mini game
     */
    private void checkOrder(IMiniAppContext miniAppContext, JSONObject params, AsyncResult result, int type) {
        Activity activity = miniAppContext.getAttachedActivity();
        if (type == PayApi.PAY_TYPE_GAME) {
            dealPayTypeGame(miniAppContext, params);
        }
        payApi.checkOrder(params, type, new PayApi.PayCallBack() {
            @Override
            public void onSuccess(JSONObject checkRet) {
                gotoPayPage(activity, miniAppContext, checkRet, result);
            }

            @Override
            public void onFailed(int errCode, String msg) {
                Log.e(TAG, "pay failed " + msg);
                activity.runOnUiThread(() -> Toast.makeText(miniAppContext.getContext(), msg, Toast.LENGTH_SHORT).show());
                result.onReceiveResult(false, new JSONObject());
            }
        });
    }

    private void gotoPayPage(Activity activity, IMiniAppContext miniAppContext, JSONObject checkRet, AsyncResult result) {
        activity.runOnUiThread(() -> {
            String fee = checkRet.optString("actualAmount");
            double paymentValue = Double.parseDouble(fee) / 10000;
            Intent intent = new Intent(activity, PaymentMethodActivity.class);
            intent.putExtra("totalFee", paymentValue);
            intent.putExtra("rawData", checkRet.toString());
            activity.startActivity(intent);
            paymentRequest = new PaymentRequest(miniAppContext, checkRet, result, paymentValue);
            Log.e("TAG", "payment request " + this.hashCode());
        });
    }

    private void dealPayTypeGame(IMiniAppContext miniAppContext, JSONObject params) {
        if (null == params) {
            Log.e(TAG, "dealPayTypeGame:ill params");
            return;
        }
        LoginApi.UserInfo userInfo = Login.g(miniAppContext.getContext()).getUserInfo();
        if (null == userInfo) {
            Log.e(TAG, "dealPayTypeGame:get user token failed");
            return;
        }
        try {
            params.put("token", userInfo.token);
            ITMFConfigManager itmfConfigManager = TMFServiceManager.getDefaultServiceManager().getService(ITMFConfigManager.class);
            String appKey = itmfConfigManager.getAppKey();
            params.put("appId", appKey);
            Log.e("TAG", "put token for game " + userInfo.token);
        } catch (JSONException ignored) {
        }
    }

    /**
     * STEP 2: show pwd dialog
     *
     * @param activity
     * @param money
     */
    private void showPwdDialog(Activity activity, double money, ICustomPayCallback payCallback, int iconSrc, String desc) {
        CustomPayDemo.CustomPayDialog customPayDialog = new CustomPayDemo.CustomPayDialog(activity, money, R.style.MyAlertDialog, iconSrc, desc);
        customPayDialog.addPayResultListen(payCallback);
        customPayDialog.show();
    }

    /**
     * STEP 3: confirm pwd input and check pwd
     *
     * @param activity
     * @param pwd
     * @param data
     * @param result
     */
    private void checkPwdAndPay(Activity activity, String pwd, JSONObject data, AsyncResult result, String payModel, String payModelId) {
        if (checkPassWord(pwd)) {
            if (isMock(activity)) {//mock payment
                requestPaymentByMock(activity, data, result);
            } else {//real payment
                requestPayment(activity, data, result, payModel, payModelId);
            }
        } else {
            JSONObject ret = new JSONObject();
            try {
                ret.put("errMsg", "bad pwd ");
            } catch (JSONException ignored) {
            }
            result.onReceiveResult(false, ret);
            activity.finish();
        }
    }

    private boolean checkPassWord(String pwd) {
        return "666666".equals(pwd);
    }


    /**
     * STEP 4: request payment
     *
     * @param activity
     * @param data
     * @param asyncResult
     */
    private void requestPayment(Activity activity, JSONObject data, AsyncResult asyncResult, String payModel, String paymdelId) {
        //real payment
        payApi.payOrder(data, new PayApi.PayCallBack() {
            @Override
            public void onSuccess(JSONObject result) {
                String totalFee = result.optString("paymentAmount");
                showPayResult(true, activity, totalFee);
                JSONObject resultToJs = result.optJSONObject("data");
                try {
                    resultToJs.put("errCode", "0");
                } catch (JSONException ignored) {
                }
                asyncResult.onReceiveResult(true, resultToJs);
            }

            @Override
            public void onFailed(int errCode, String msg) {
                JSONObject ret = new JSONObject();
                try {
                    ret.put("errMsg", msg);
                    ret.put("errCode", errCode);
                } catch (JSONException ignored) {
                }
                asyncResult.onReceiveResult(false, ret);
                activity.finish();

            }
        }, payModel, paymdelId);
    }


    private void checkOrderMock(IMiniAppContext miniAppContext, JSONObject params, AsyncResult result) {
        Activity activity = miniAppContext.getAttachedActivity();
        String fee = params.optString("total_fee");
        double paymentValue = Double.parseDouble(fee) / 10000;
        showPwdDialog(activity, paymentValue, (retCode, msg, dialogInterface) -> {
            String pwd = ((CustomPayDemo.CustomPayDialog) dialogInterface).getInputText();
            Log.e(TAG, "onDismiss isComplete=" + pwd);
            if (!TextUtils.isEmpty(pwd)) {
                checkPwdAndPay(activity, pwd, params, result, "wechat", "wechat pay");
            } else {
                Log.e(TAG, "empty pwd ~");
                result.onReceiveResult(false, new JSONObject());
            }
        }, R.drawable.pay_method_wechat, miniAppContext.getContext().getString(R.string.tcmpp_payment_we_chat_pay));
    }

    private void requestPaymentByMock(Activity activity, JSONObject data, AsyncResult asyncResult) {
        String totalFee = data.optString("total_fee");
        showPayResult(true, activity, totalFee);
        asyncResult.onReceiveResult(true, new JSONObject());
    }

    private boolean isMock(Context context) {
        return GlobalConfigureUtil.getGlobalConfig(context).mockApi;
    }

    private void showPayResult(boolean success, Activity activity, String total) {
        Intent intent = new Intent(activity, PaymentResultActivity.class);
        intent.putExtra("success", success);
        intent.putExtra("total", total);
        activity.startActivity(intent);
        activity.finish();
    }

    private static class PaymentRequest {
        public IMiniAppContext miniAppContext;
        public JSONObject params;
        public AsyncResult result;
        public double payValue;

        public PaymentRequest(IMiniAppContext miniAppContext, JSONObject params, AsyncResult result, double payValue) {
            this.miniAppContext = miniAppContext;
            this.params = params;
            this.result = result;
            this.payValue = payValue;

        }
    }
}

File: Kotlin/app/src/main/java/com/tencent/tcmpp/demo/open/payment/PaymentManager.kt
package com.tencent.tcmpp.demo.open.payment

import android.app.Activity
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.text.TextUtils
import android.util.Log
import android.widget.Toast
import com.tencent.tcmpp.demo.R
import com.tencent.tcmpp.demo.activity.PaymentResultActivity
import com.tencent.tcmpp.demo.open.login.Login
import com.tencent.tcmpp.demo.utils.GlobalConfigureUtil
import com.tencent.tmf.base.api.config.ITMFConfigManager
import com.tencent.tmf.core.api.TMFServiceManager
import com.tencent.tmfmini.sdk.launcher.core.IMiniAppContext
import com.tencent.tmfmini.sdk.launcher.core.proxy.AsyncResult
import org.json.JSONException
import org.json.JSONObject

class PaymentManager private constructor() {

    companion object {
        private const val TAG = "PaymentManager"

        @Volatile
        private var instance: PaymentManager? = null

        fun g(): PaymentManager {
            return instance ?: synchronized(this) {
                instance ?: PaymentManager().also { instance = it }
            }
        }
    }

    private val payApi = PayApi()
    private var paymentRequest: PaymentRequest? = null

    fun startPayment(miniAppContext: IMiniAppContext, params: JSONObject, result: AsyncResult) {
        // do params check
        if (!params.has("prepayId")) {
            // ill params
            val failRet = JSONObject().apply {
                try {
                    put("success", false)
                } catch (ignored: JSONException) {
                }
            }
            result.onReceiveResult(false, failRet)
            return
        }

        // mock check order
        if (isMock(miniAppContext.context)) {
            val mockData = JSONObject().apply {
                try {
                    put("total_fee", 9999)
                } catch (ignored: JSONException) {
                }
            }
            checkOrderMock(miniAppContext, mockData, result)
            return
        }

        // real check order
        checkOrder(miniAppContext, params, result, PayApi.PAY_TYPE_APP)
    }

    fun notifyPaymentCancel() {
        paymentRequest?.let { request ->
            val ret = JSONObject().apply {
                try {
                    put("errMsg", "cancel")
                } catch (ignored: JSONException) {
                }
            }
            request.result.onReceiveResult(false, ret)
            paymentRequest = null
        }
    }

    fun showPwdConfirm(activity: Activity, iconSrc: Int, desc: String, model: String, modelId: String) {
        paymentRequest?.let { request ->
            val result = request.result
            val checkRet = request.params
            try {
                checkRet.put("appId", request.miniAppContext.miniAppInfo.appId)
            } catch (ignored: JSONException) {
            } catch (ignored: NullPointerException) {
            }

            showPwdDialog(activity, request.payValue, object : ICustomPayCallback {
                override fun onPayResult(retCode: Int, msg: String, dialogInterface: DialogInterface?) {
                    if (retCode == 1) {
                        val pwd = (dialogInterface as CustomPayDemo.CustomPayDialog).inputText
                        Log.e(TAG, "onDismiss isComplete=$pwd")
                        if (!TextUtils.isEmpty(pwd)) {
                            checkPwdAndPay(activity, pwd, checkRet, result, model, modelId)
                        } else {
                            Log.e(TAG, "empty pwd ~")
                            val ret = JSONObject().apply {
                                try {
                                    put("errMsg", "empty pws")
                                } catch (ignored: JSONException) {
                                }
                            }
                            result.onReceiveResult(false, ret)
                            activity.finish()
                        }
                        paymentRequest = null
                    }
                }
            }, iconSrc, desc)
        }
    }

    /**
     * STEP 1: check order status
     */
    private fun checkOrder(miniAppContext: IMiniAppContext, params: JSONObject, result: AsyncResult, type: Int) {
        val activity = miniAppContext.attachedActivity
        if (type == PayApi.PAY_TYPE_GAME) {
            dealPayTypeGame(miniAppContext, params)
        }

        payApi.checkOrder(params, type, object : PayApi.PayCallBack {
            override fun onSuccess(checkRet: JSONObject) {
                gotoPayPage(activity, miniAppContext, checkRet, result)
            }

            override fun onFailed(errCode: Int, msg: String) {
                Log.e(TAG, "pay failed $msg")
                activity.runOnUiThread {
                    Toast.makeText(miniAppContext.context, msg, Toast.LENGTH_SHORT).show()
                }
                result.onReceiveResult(false, JSONObject())
            }
        })
    }

    private fun gotoPayPage(activity: Activity, miniAppContext: IMiniAppContext, checkRet: JSONObject, result: AsyncResult) {
        activity.runOnUiThread {
            val fee = checkRet.optString("actualAmount")
            val paymentValue = fee.toDouble() / 10000
            val intent = Intent(activity, PaymentMethodActivity::class.java).apply {
                putExtra("totalFee", paymentValue)
                putExtra("rawData", checkRet.toString())
            }
            activity.startActivity(intent)
            paymentRequest = PaymentRequest(miniAppContext, checkRet, result, paymentValue)
            Log.e("TAG", "payment request ${this.hashCode()}")
        }
    }

    private fun dealPayTypeGame(miniAppContext: IMiniAppContext, params: JSONObject?) {
        if (params == null) {
            Log.e(TAG, "dealPayTypeGame:ill params")
            return
        }

        val userInfo = Login.g(miniAppContext.context).getUserInfo()
        if (userInfo == null) {
            Log.e(TAG, "dealPayTypeGame:get user token failed")
            return
        }

        try {
            params.put("token", userInfo.token)
            val itmfConfigManager = TMFServiceManager.getDefaultServiceManager().getService(ITMFConfigManager::class.java)
            val appKey = itmfConfigManager.appKey
            params.put("appId", appKey)
            Log.e("TAG", "put token for game ${userInfo.token}")
        } catch (ignored: JSONException) {
        }
    }

    /**
     * STEP 2: show pwd dialog
     */
    private fun showPwdDialog(activity: Activity, money: Double, payCallback: ICustomPayCallback, iconSrc: Int, desc: String) {
        val customPayDialog = CustomPayDemo.CustomPayDialog(activity, money, R.style.MyAlertDialog, iconSrc, desc)
        customPayDialog.addPayResultListen(payCallback)
        customPayDialog.show()
    }

    /**
     * STEP 3: confirm pwd input and check pwd
     */
    private fun checkPwdAndPay(activity: Activity, pwd: String, data: JSONObject, result: AsyncResult, payModel: String, payModelId: String) {
        if (checkPassWord(pwd)) {
            if (isMock(activity)) { // mock payment
                requestPaymentByMock(activity, data, result)
            } else { // real payment
                requestPayment(activity, data, result, payModel, payModelId)
            }
        } else {
            val ret = JSONObject().apply {
                try {
                    put("errMsg", "bad pwd ")
                } catch (ignored: JSONException) {
                }
            }
            result.onReceiveResult(false, ret)
            activity.finish()
        }
    }

    private fun checkPassWord(pwd: String): Boolean {
        return "666666" == pwd
    }

    /**
     * STEP 4: request payment
     */
    private fun requestPayment(activity: Activity, data: JSONObject, asyncResult: AsyncResult, payModel: String, payModelId: String) {
        // real payment
        payApi.payOrder(data, object : PayApi.PayCallBack {
            override fun onSuccess(result: JSONObject) {
                val totalFee = result.optString("paymentAmount")
                showPayResult(true, activity, totalFee)
                val resultToJs = result.optJSONObject("data")
                try {
                    resultToJs?.put("errCode", "0")
                } catch (ignored: JSONException) {
                }
                asyncResult.onReceiveResult(true, resultToJs ?: JSONObject())
            }

            override fun onFailed(errCode: Int, msg: String) {
                val ret = JSONObject().apply {
                    try {
                        put("errMsg", msg)
                        put("errCode", errCode)
                    } catch (ignored: JSONException) {
                    }
                }
                asyncResult.onReceiveResult(false, ret)
                activity.finish()
            }
        }, payModel, payModelId)
    }

    private fun checkOrderMock(miniAppContext: IMiniAppContext, params: JSONObject, result: AsyncResult) {
        val activity = miniAppContext.attachedActivity
        val fee = params.optString("total_fee")
        val paymentValue = fee.toDouble() / 10000

        showPwdDialog(activity, paymentValue, object : ICustomPayCallback {
            override fun onPayResult(retCode: Int, msg: String, dialogInterface: DialogInterface?) {
                val pwd = (dialogInterface as CustomPayDemo.CustomPayDialog).inputText
                Log.e(TAG, "onDismiss isComplete=$pwd")
                if (!TextUtils.isEmpty(pwd)) {
                    checkPwdAndPay(activity, pwd, params, result, "wechat", "wechat pay")
                } else {
                    Log.e(TAG, "empty pwd ~")
                    result.onReceiveResult(false, JSONObject())
                }
            }
        }, R.drawable.pay_method_wechat, miniAppContext.context.getString(R.string.tcmpp_payment_we_chat_pay))
    }

    private fun requestPaymentByMock(activity: Activity, data: JSONObject, asyncResult: AsyncResult) {
        val totalFee = data.optString("total_fee")
        showPayResult(true, activity, totalFee)
        asyncResult.onReceiveResult(true, JSONObject())
    }

    private fun isMock(context: Context): Boolean {
        return GlobalConfigureUtil.getGlobalConfig(context).mockApi
    }

    private fun showPayResult(success: Boolean, activity: Activity, total: String) {
        val intent = Intent(activity, PaymentResultActivity::class.java).apply {
            putExtra("success", success)
            putExtra("total", total)
        }
        activity.startActivity(intent)
        activity.finish()
    }

    private data class PaymentRequest(
        val miniAppContext: IMiniAppContext,
        val params: JSONObject,
        val result: AsyncResult,
        val payValue: Double
    )
}